### PR TITLE
executor: speed up replace into statement

### PR DESCRIPTION
**Pull request by jackysp**
## What have you changed? (mandatory)
Speed up `replace into` statement.

## What are the type of the changes (mandatory)?
- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested (mandatory)?
Unit tests and manual tests.

## Benchmark result if necessary (optional)
```
[17:13:29] yusp@ShuaipengdeMacBook-Pro ~/cluster
$ time mysql -uroot -P4000 -h127.0.0.1 --default-character-set utf8 -D test < replace.sql 

real	0m1.964s
user	0m0.023s
sys	0m0.015s
[17:13:40] yusp@ShuaipengdeMacBook-Pro ~/cluster
$ time mysql -uroot -P5000 -h127.0.0.1 --default-character-set utf8 -D test < replace.sql 

real	0m19.322s
user	0m0.025s
sys	0m0.020s
```
This pr(port 4000) is nearly 10x faster than master(port 5000).

PTAL @coocood @zimulala

### CODE REVIEW

File: executor/batch_checker.go [hunk 1]
```diff
@@ -249,3 +249,29 @@ func (b *batchChecker) deleteDupKeys(row toBeCheckedRow) {
 		delete(b.dupKVs, string(uk.newKV.key))
 	}
 }
+
+// decodeOldRow decodes the table record row from storage for batch check.
+func (b *batchChecker) decodeOldRow(ctx sessionctx.Context, t table.Table, handle int64) (types.DatumRow, error) {
```
[the remark this answers is not in the document]
Comment (zz-jason) [reply executor/batch_checker.go]: vote for `getOldRow` 😄

**Comment by zz-jason** [comment]
@jackysp PTAL the comments

**Comment by jackysp** [comment]
PTAL @coocood @shenli @zz-jason

**Comment by zz-jason** [comment]
LGTM

**Comment by jackysp** [comment]
PTAL @coocood

File: executor/replace.go [hunk 2]
```diff
@@ -57,46 +108,67 @@ func (e *ReplaceExec) exec(rows []types.DatumRow) error {
...
+	}
+
+	// Batch get the to-be-replaced rows in storage.
+	err = e.initDupOldRowValue(e.ctx, e.Table, newRows)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	for i, r := range e.toBeCheckedRows {
+		for {
```
Comment (coocood) [inline executor/replace.go]: Can we extract the inner for loop into its own function.
It's hard to see where the code is going to execute after `break` and `continute`.

**Comment by jackysp** [comment]
PTAL @coocood

**Comment by jackysp** [comment]
PTAL @coocood

File: executor/replace.go [hunk 3]
```diff
@@ -44,7 +46,122 @@ func (e *ReplaceExec) Open(ctx context.Context) error {
...
+	if err != nil {
+		return 0, errors.Trace(err)
+	}
+	if !e.ctx.GetSessionVars().ImportingData {
+		e.ctx.StmtAddDirtyTableOP(DirtyTableAddRow, e.Table.Meta().ID, h, row)
+	}
+	return h, nil
+}
+
+// replaceRow remove all duplicate rows for one row, then insert it.
```
Comment (zimulala) [inline executor/replace.go]: s/remove/removes
s/insert/inserts

**Comment by jackysp** [comment]
PTAL @zimulala

**Comment by coocood** [comment]
LGTM

**Pull request merged**